### ISSUE TITLE
PHPStan complains about return type of instantiateWith in a PersistentProxyObjectFactory factory

### ISSUE THREAD

**Issue by marien-probesys** [zenstruck/foundry#721]
Hi,

I use PHPStan to verify the typing in my application. I have an issue when using the `instantiateWith()` method in a `PersistentProxyObjectFactory` factory.

`instantiateWith()` is declared in the `ObjectFactory` and must return an object of the template class `T`. However, when inheriting from the `PersistentProxyObjectFactory` class, `T` becomes [`T&Proxy<T>`](https://github.com/zenstruck/foundry/blob/2.x/src/Persistence/PersistentProxyObjectFactory.php#L21C37-L21C47). Meaning that PHPStan expects `instantiateWith()` to return a `T&Proxy<T>` object.

```php
namespace App\Tests\Factory;

use App\Entity\MyEntity;
use Zenstruck\Foundry\Persistence\PersistentProxyObjectFactory;

/**
 * @extends PersistentProxyObjectFactory<MyEntity>
 */
final class MyEntityFactory extends PersistentProxyObjectFactory
{
    protected function initialize(): static
    {
        // The return type of instantiateWith is correct, but PHPStan expects a MyEntity&Proxy<MyEntity> and fail.
        return $this->instantiateWith(function (array $attributes, string $class): MyEntity {
            $myEntity = new MyEntity();

            // …

            return $myEntity;
        });
    }

    // …
}
```

**Comment by nikophil** [comment]
Hi @marien-probesys 

which Foundry's (and PHPStan) version are you using?
I've recently added [specific phpdoc](https://github.com/zenstruck/foundry/blob/2.x/src/ObjectFactory.php#L71) for this case

**Comment by marien-probesys** [comment]
Hi @nikophil, thanks for your answer. I'm using PHPStan 1.12.11 (the latest version of v1) and Foundry 2.2.2.

I'm sorry I made a mistake in my explanations: the problem is not with the `instantiateWith()` method directly, but with its callable parameter `$instantiator`. It is its return type which is wrong! (so the `InstantiatorCallable` type defined at the top of the file)

If you need it, I can create a minimal project tomorrow in order to reproduce the problem.

**Comment by marien-probesys** [comment]
I've created a repository to reproduce the problem: https://github.com/marien-probesys/foundry-721

You can see the error here: https://github.com/marien-probesys/foundry-721/actions/runs/11915667680/job/33206464561

And the factory: https://github.com/marien-probesys/foundry-721/blob/main/src/Factory/PostFactory.php

**Comment by nikophil** [comment]
thanks for your reproducer! I'll soon check this

**Comment by nikophil** [comment]
Hi, I really don't know what to do to make PHPStan happy here :thinking: 

you code seems OK, but [the phpdoc for `ObjectFactory::instantiateWith()`](https://github.com/zenstruck/foundry/blob/2.x/src/ObjectFactory.php#L22) seems also OK :thinking: 

@kbond any thoughts?

**Comment by kbond** [comment]
That's a bummer. Feels like it's because the string in the closure isn't being considered `class-string<Post>`?

**Comment by nikophil** [comment]
nope, it is because of the proxy: if the factory extends from `PersistentObjectFactory` instead of `PersistentProxyObjectFactory`, phpstan is happy

**Comment by nikophil** [comment]
I'm closing this issue. The problem is coming from the proxy mechanism which will be dropped with PHP 8.4 and lazy proxies
see https://github.com/zenstruck/foundry/pull/943
